### PR TITLE
Make `seed` reach dataloader workers so augmentations vary between runs

### PR DESCRIPTION
**Pull request by yentur**
`seed` never reaches the dataloader workers, so every value of `seed` trains on the same augmentation stream.

`build_dataloader` seeds its `torch.Generator` from a fixed constant:

```python
generator.manual_seed(6148914691236517205 + RANK)
```

Workers take their `numpy` and `random` seeds from `seed_worker`, which reads `torch.initial_seed()`, and that derives from this generator. The constant never changes, so the shuffle order and every augmentation drawn inside a worker are the same no matter what `seed` is. With `workers=0` the augmentations run in the main process, which `init_seeds(args.seed + 1 + RANK)` has already seeded, and there `seed` does take effect. The two paths disagree.

### Reproduction

SHA-256 over `img` and `bboxes` for two epochs of coco8 at `imgsz=64`, after `init_seeds(seed + 1)` the way `BaseTrainer` does it:

```
workers=0
  seed=0      11de60115ffa2390
  seed=1      7a4ad2d233f229db
  seed=12345  dfb63119683c77e2

workers=2
  seed=0      4dc0eb59ef9a8438
  seed=1      4dc0eb59ef9a8438
  seed=12345  4dc0eb59ef9a8438
```

`workers` defaults to 8, so the second block is what training actually does.

After this change:

```
workers=2
  seed=0      4dc0eb59ef9a8438
  seed=1      0a4e8a334886103a
  seed=12345  5f6b59a9c4356956
```

### Change

`build_dataloader` gains a `seed` argument defaulting to 0 and mixes it into the generator. `DetectionTrainer.get_dataloader` and `ClassificationTrainer.get_dataloader` pass `self.args.seed`. Segment, pose, obb, world and yoloe trainers inherit the detection method. Validators are left alone since they neither shuffle nor augment.

### Effect on existing runs

`6148914691236517205 + RANK + 0` is the old constant and 0 is the default, so default training is bit identical. The `seed=0` hash above is the same before and after. Runs that set `seed` to a non-zero value will now get a different shuffle order and different augmentations, and their numbers will move. That is the behavior the argument is supposed to have, but it is a change for those users.

### Test

`test_dataloader_seed_varies_sampling_order` builds three loaders with seeds 0, 0 and 1, then checks the first two agree and the third does not. Reverting only the generator line:

```
>           assert first != other  # different seeds must not share one order
E           assert [47, 15, 58, 46, 5, 2, ...] != [47, 15, 58, 46, 5, 2, ...]
1 failed in 0.20s
```

Restored, it passes. Also ran `pytest tests/test_python.py -k "dataloader or cfg or seed"` (6 passed) and one-epoch CPU trainings of `yolo11n.yaml` on coco8 and `yolo11n-cls.yaml` on imagenet10 with `seed=7, workers=2`, both of which complete.

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Ensure the training seed reaches PyTorch dataloaders so sampling order and worker-based augmentations vary correctly between runs.

### 📊 Key Changes
- Added a `seed` parameter to `build_dataloader()` and incorporated it into the dataloader generator seed.
- Passed `self.args.seed` through classification and detection dataloaders.
- Added a regression test confirming identical seeds reproduce sampling order while different seeds produce different orders.

### 🎯 Purpose & Impact
- Fixes reproducibility behavior where different run seeds could replay the same sampling and augmentation sequence.
- Preserves deterministic results for matching seeds while enabling variation across distinct seeds.
- Improves training reliability for classification and object detection workflows with minimal API changes.

### CODE REVIEW

**Comment by yentur** [comment]
I have read the CLA Document and I sign the CLA

**Comment by UltralyticsAssistant** [comment]
👋 Hello @yentur, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

This automated review has been posted for initial guidance, and an engineer will assist with the next steps. 🛠️

File: ultralytics/data/build.py [hunk 1]
```diff
@@ -362,7 +365,9 @@ def build_dataloader(
     # persistent DataLoader worker pools that add overhead and can stall tiny datasets while holding CUDA context.
     nw = min(os.cpu_count() // max(nd, 1), workers, 0 if batches <= 1 else batches)  # number of workers
     generator = torch.Generator()
-    generator.manual_seed(6148914691236517205 + RANK)
+    # Workers derive their numpy/random seeds from this generator via seed_worker, so `seed` has to reach it for
+    # augmentations to vary between runs. Without it every seed replays one fixed augmentation stream.
+    generator.manual_seed(6148914691236517205 + RANK + seed)
```
Comment (UltralyticsAssistant) [inline ultralytics/data/build.py]: 💡 **MEDIUM**: This only seeds `DataLoader`'s generator. In distributed training, `sampler` is a `DistributedSampler`, whose shuffle sequence comes from its own `seed` argument (currently left at its default of `0` at line 355); `BaseTrainer` then only varies that sequence with `set_epoch(epoch)`. Consequently, different `args.seed` values still use identical per-epoch sampling orders under DDP, despite the new API documentation promising seed-dependent sampling. Pass the run seed to `DistributedSampler` as well (identically on every rank).

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change correctly propagates the configured seed to the DataLoader generator, which fixes worker RNG seeding and single-process shuffled sampling. DDP sampling remains seed-independent because the distributed sampler is not seeded from the new argument.

💬 Posted 1 inline comment

**Review by glenn-jocher — APPROVED** [review]
LGTM. Personally reviewed and reproduced the exact head, with an independent cold full-diff review. The dataloader owner now derives one common run seed for all global DDP ranks without adding a public/default argument, keeps rank-distinct worker streams, preserves seed-0 behavior, and safely wraps the generator mix. Single-process, 2-node/4-rank, max-uint64, CI, mergeability, and review-thread gates are all clean.
